### ISSUE TITLE
LA+UU combination doesn't allocate enough space in Noto Sans Devanagari and UI

### ISSUE THREAD

**Issue by marekjez86** [googlei18n/noto-fonts#410]
The LA+UU combination doesn't allocate enough space to the UU. This is visible when you try to render a string that has a punctuation immediately after the word. I'm attaching screenshots of HarfBuzz rendering the string "चालू; चालू/ चालू, (चालू) [चालू]" with the latest Noto Sans Devanagari fonts. You can see that the punctuation intersects the UU.
![noto-ui](https://cloud.githubusercontent.com/assets/13322623/8737316/905eb6fe-2bd5-11e5-8c01-42f0bfac18fc.png)
![noto](https://cloud.githubusercontent.com/assets/13322623/8737317/905f7c24-2bd5-11e5-98c8-a18469c0d6eb.png)

**Comment by schriftgestalt** [comment]
That is a kerning problem. One would need to add contextual kerning for the string: 

    [all Fullforms with a stem on the right] UU slash 
    [all Fullforms with a stem on the right] UU comma

if you just kern UU+fraction, the combination with ka+UU would be to wide.

**Comment by fontguy** [comment]
I haven't seen any other Indic fonts that do this. This is a great feature to add but I'm not sure that it is a "High Priority" item.

**Comment by pauldhunt** [comment]
you don’t necessarily need contextual kerning, all you need is a set of rules that do not ignore marks, then you can kern things like UUsign @punctuation.

**Comment by schriftgestalt** [comment]
> you don’t necessarily need contextual kerning, all you need is a set of rules that do not ignore marks, then you can kern things like: UUsign + @punctuation.

That would not work. As the UUsign is a mark and the kerning should only be applied if the UU is attached to a right stem. There are other letters that position the UU more to the left (KA). Then you don't need the kerning.

**Comment by JelleBosmaMT** [comment]
The context rule would have to be much more complicated. There might be a reph, anusvara, candrabindu, one of the vedic signs and combinations thereof. And of course, there are other marks and mark combinations which may collide too. The spacing of the punctuation is set to be generous enough that issues of legibility should be limited to very rare combinations of marks.
As the designer of the font I would love to go into this kind of details. But the design of a Devanagari font is an infinite amount of work, which I am expected to do in a very finite amount of time. There are about a million other details which I would like to improve given the time. In real world usage of Noto Sans Devanagari, the punctuation marks are usually taken from the Latin font (Android's Roboto font). So in a priority list, this is pretty close to the bottom.

**Comment by JelleBosmaMT** [comment]
Secondly, one has to consider in as far this kerning would solve a problem.We would exchange touching characters by extra white space. In particular: if you use the slash to separate two words, there is a choice between touching the vowel and having the slash much closer to the second word than to the first. The gap of the solution could be worse than the touch of the problem.

**Comment by xiangyexiao** [comment]
@JelleBosmaMT Does Phase III Deva fonts behave the same as Phase II fonts, or have improvement on the above mentioned UU+punctuation pairs?

**Comment by JelleBosmaMT** [comment]
Yes. This is non-issue. Given that the tail touches at near perpendicular angle, there is no problem with legibility or esthetics.

**Comment by xiangyexiao** [comment]
This is not a Noto font issue. This needs to be improved to improve font picking algorithm on Android side. @roozbehp

**Comment by xiangyexiao** [comment]
I will keep it open and wait for other people to comment.

**Comment by roozbehp** [comment]
The bug is perpendicular to the font selection algorithm on Android. I understand @JelleBosmaMT is not happy with the current font selection algorithm in Android, but we are constantly improving that algorithm and moving towards the direction he likes. I cannot share more specific details publicly, unfortunately.

As for this specific bug, the problem seems to arise from a combination of the UU vowel intruding its box and lack of kerning with certain characters that may appear to its right. The technical solution may be either adjusting the advance of UU or kerning them. To be honest, it doesn't matter that much to us, as far as the text looks good, and in this case, it seems that it doesn't look good.

/cc @raphlinus @nona-google

**Comment by JelleBosmaMT** [comment]
Changing the advance of the non-spacing mark UU would mean that we introduce a critical bug as a technical solution for something that is not a problem ;-(

**Comment by nizarsq** [comment]
No changes noticed in allocating  more space for both NotoSansDevanagari or NotoSansDevanagariUI
<img width="1106" alt="Screen Shot 2020-07-31 at 2 05 12 PM" src="https://user-images.githubusercontent.com/22158882/89077352-dac10900-d336-11ea-8fe0-c202466a8788.png">

<img width="1016" alt="Screen Shot 2020-07-31 at 2 03 57 PM" src="https://user-images.githubusercontent.com/22158882/89077267-af3e1e80-d336-11ea-8833-2e0bc3ddf1e8.png">

**Comment by simoncozens** [comment]
A note about other Devanagari fonts:

* Many do nothing, and there are clashes in all these cases. (Example: Kohinoor Devanagari, Mukta, Hind.)
* Some avoid the problem by having generous sidebearings on the punctuation. (Devanagari MT and ITF Devanagari) - not appropriate in our case.
* Some (e.g. Poppins) avoid it just via sheer proportions.
* Tiro Devanagari is the only one I've found which *actively* works around it, adding kerning to the uu mark before comma and semicolon, but not before slash or brackets.
